### PR TITLE
Implement client reset with automatic recovery

### LINKED ISSUE #2745: Implement Client Reset w/ Recovery

### PR DESCRIPTION
**Pull request by LaPeste**
## Description

Fixes #2745 

##  TODO

* [x] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

File: Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryOrDiscardLocalHandler.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    /// while the last one will be invoked in case an error occurs during the automated process and the system needs to fallback to a manual mode.
+    /// The overall recommendation for using this strategy is that using the three available callbacks should only be considered when:
+    /// 1. the user needs to be notified (in <see cref="OnBeforeReset"/>) of an incoming potential data loss of unsynced data as a result of a merge
+    ///    or a complete discard of local changes
+    /// 2. the user needs to be notified (in <see cref="OnAfterReset"/>) that the reset process has completed
+    /// 3. advanced use cases for data-sensitive applications where the developer wants to recover in the most appropriate way the unsynced data
+    /// 4. backup the whole realm before the client reset happens (in <see cref="OnBeforeReset"/>). Such backup could, for example, be used to restore the unsynced data (see 3.)
+    /// </remarks>
+    /// <seealso href="https://docs.mongodb.com/realm/sdk/dotnet/advanced-guides/client-reset/">Client Resets - .NET SDK</seealso>
+    public sealed class AutomaticRecoveryOrDiscardLocalHandler : ClientResetHandlerBase
```
Comment (nirinchev) [inline Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryOrDiscardLocalHandler.cs:41]: Let's not have two of those (especially with such an obnoxious name). Instead we can add an optional ctor argument that controls the fallback:

```csharp
public AutomaticReccoveryHandler(Fallback fallback = Fallback.DiscardLocal)
```

File: Realm/Realm/Configurations/SyncConfigurationBase.cs [hunk 2]
```diff
@@ -133,11 +134,18 @@ internal override async Task<SharedRealmHandle> CreateHandleAsync(Schema.RealmSc
...
         internal virtual Native.SyncConfiguration CreateNativeSyncConfiguration()
         {
-            var clientResyncMode = ClientResyncMode.DiscardLocal;
+            var clientResyncMode = ClientResyncMode.AutomaticRecoveryOrDiscardLocal;
 
-            if (ClientResetHandler != null && ClientResetHandler is ManualRecoveryHandler)
+            if (ClientResetHandler != null)
             {
-                clientResyncMode = ClientResyncMode.Manual;
+                clientResyncMode = ClientResetHandler switch
```
Comment (nirinchev) [inline Realm/Realm/Configurations/SyncConfigurationBase.cs:141]: Instead of switching on type, can we add an abstract internal property to the ClientResetHandlerBase class and have inheritors be explicit about the resync mode they support.

**Review by nirinchev — CHANGES_REQUESTED** [review]
Haven't reviewed the tests, but have some comments about the implementation.

File: Realm/Realm/Configurations/SyncConfigurationBase.cs (hunk 2 above)
Comment (LaPeste) [reply Realm/Realm/Configurations/SyncConfigurationBase.cs:141]: Beside allowing a param to the constructor of `AutomaticReccoveryHandler`, as you suggested [here](https://github.com/realm/realm-dotnet/pull/2956#discussion_r899977885). Is there any other advantage to this approach? E.g. do you see handlers ever changing internal behaviour after being released? Also reading a property is surely faster than switching on the type, but were you concerned with performance?

Comment (LaPeste) [reply Realm/Realm/Configurations/SyncConfigurationBase.cs:141]: All I'm trying to say is, the type of the recovery strategy already implies what the recovery behaviour is gonna be. Why do we need to write it "again" in the class?

File: Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+namespace Realms.Sync.ErrorHandling
+{
+    // TODO andrea: ask how this strategy may fail if it isn't for the server not allowing automatic merge
```
Comment (LaPeste) [inline Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryHandler.cs:21]: I'm waiting on James on this one.

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 4]
```diff
@@ -39,6 +42,13 @@ public class SessionTests : SyncTestBase
     {
         private readonly ConcurrentQueue<EventHandler<ErrorEventArgs>> _sessionErrorHandlers = new();
 
+        private static IEnumerable<IClientResetHandler> _allClientResetHandlers()
+        {
+            yield return new ClientResetHandler<DiscardLocalResetHandler>();
+            yield return new ClientResetHandler<AutomaticRecoveryHandler>(AutomaticRecoveryHandler.Fallback.DiscardLocal);
+            yield return new ClientResetHandler<AutomaticRecoveryHandler>(AutomaticRecoveryHandler.Fallback.None);
+        }
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:50]: This seems a bit overengineered and will not work for Unity. When passing arguments to `TestCaseSource`, you need to have them be `object[]`. My suggestion would be to do something like:

```csharp
public static readonly object[] AllClientResetHandlers = new[]
{
    new object[] { typeof(DiscardLocalHandler), (AutomaticRecoveryHandler.Fallback?)null },
    new object[] { typeof(AutomaticRecoveryHandler), AutomaticRecoveryHandler.Fallback.DiscardLocal },
    new object[] { typeof(AutomaticRecoveryHandler), AutomaticRecoveryHandler.Fallback.None },
};

[TestCaseSource(nameof(AllClientResetHandlers))]
public void MyTest(Type handlerType, AutomaticRecoveryHandler.Fallback? fallback)
{

}

private static ClientResetHandlerBase GetClientResetHandler(Type type, 
    AutomaticRecoveryHandler.Fallback? fallback,
    BeforeResetCallback beforeCb = null,
    AfterResetCallback afterCb = null,
    ClientResetCallback manualCb = null)
{
    var handler = fallback.HasValue ? Activator.CreateInstance(type, new[] { fallback.Value }) : Activator.CreateInstance(type);

    if (beforeCb != null)
    {
        type.GetProperty(nameof(DiscardLocalHandler. OnBeforeReset)).SetValue(handler, beforeCb);
    }

    // rest of callbacks
}
```

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 5]
```diff
@@ -270,31 +237,388 @@ public void Session_ClientReset_DiscardLocal_ManualResetFallback_InitiateClientR
...
+        public void SessionIntegrationTest_ClientResetHandlers_OutOfBoundArrayMove_Discarded()
+        {
+            SyncTestHelpers.RunBaasTestAsync(async () =>
+            {
+                var user = await GetUserAsync();
+                var partition = Guid.NewGuid().ToString();
+
+                // ===== clientA =====
+                var tscAfterClientResetMergeA = new TaskCompletionSource<object>();
+                var configA = await GetIntegrationConfigAsync(partition, null, "realm123.realm", user);
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:297]: It's a bad practice to pass `null` as a positional argument as it's not clear what it means. Since that is an optional argument, it's a better idea to name the remaining arguments. Also, let's not specify the path of the realm explicitly.

```suggestion
                var configA = await GetIntegrationConfigAsync(partition, user: user);
```

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 6]
```diff
@@ -270,31 +237,388 @@ public void Session_ClientReset_DiscardLocal_ManualResetFallback_InitiateClientR
...
+                realmA.Write(() =>
+                {
+                    originalObj.Strings.RemoveAt(2);
+                });
 
-                        onBeforeTriggered = true;
-                        tcs.TrySetResult(null);
-                    })
+                // ===== clientB =====
+                var configB = await GetIntegrationConfigAsync(partition, null, "realm456.realm", user);
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:335]:
```suggestion
                var configB = await GetIntegrationConfigAsync(partition, optionalPath: configA.DatabasePath + "_B", user: user);
```

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 7]
```diff
@@ -270,31 +237,388 @@ public void Session_ClientReset_DiscardLocal_ManualResetFallback_InitiateClientR
...
+                var afterCbB = GetOnAfterHandler(tcsAfterClientResetB, (before, after) =>
+                {
+                    var listB = after.All<SyncObjectWithRequiredStringList>().Single().Strings.ToList();
+                    var list = after.All<SyncObjectWithRequiredStringList>().Single().Strings;
+                    Assert.That(list.Count, Is.EqualTo(3));
+                    Assert.That(list[0], Is.EqualTo("2"));
+                    Assert.That(list[1], Is.EqualTo("0"));
+                    Assert.That(list[2], Is.EqualTo("1"));
+                });
+                configB.ClientResetHandler = new AutomaticRecoveryHandler(AutomaticRecoveryHandler.Fallback.DiscardLocal)
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:347]:
```suggestion
                configB.ClientResetHandler = new AutomaticRecoveryHandler
```

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 8]
```diff
@@ -270,31 +237,388 @@ public void Session_ClientReset_DiscardLocal_ManualResetFallback_InitiateClientR
...
+                var realmB = await GetRealmAsync(configB);
+
+                await WaitForDownloadAsync(realmB);
+
+                Assert.That(realmB.All<SyncObjectWithRequiredStringList>().Count, Is.EqualTo(1));
+                var originalObjStr = realmB.All<SyncObjectWithRequiredStringList>().Single().Strings;
+                Assert.That(originalObjStr.Count, Is.EqualTo(3));
+                Assert.That(originalObjStr.First(), Is.EqualTo("0"));
+                Assert.That(originalObjStr.ElementAt(1), Is.EqualTo("1"));
+                Assert.That(originalObjStr.ElementAt(2), Is.EqualTo("2"));
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:360]:
```suggestion
                Assert.That(originalObjStr, Is.EqualTo(new[] { "0", "1", "2"}));
```

**Review by nirinchev — CHANGES_REQUESTED** [review]

File: Tests/Tests.iOS/Tests.iOS.csproj [hunk 9]
```diff
@@ -30,6 +30,8 @@
     <MtouchLink>SdkOnly</MtouchLink>
     <MtouchDebug>true</MtouchDebug>
     <MtouchExtraArgs>--nosymbolstrip --linkskip=nunit.framework --linkskip=Realm.Tests --nolinkaway --linkskip=Newtonsoft.Json</MtouchExtraArgs>
+    <CodesignProvision></CodesignProvision>
+    <CodesignKey>iPhone Developer</CodesignKey>
```
[the remark this answers is not in the document]
Comment (LaPeste) [reply Tests/Tests.iOS/Tests.iOS.csproj:34]: I don't really know how I didn't see this. Thanks.

File: Tests/Realm.Tests/Sync/SessionTests.cs (hunk 4 above)
Comment (LaPeste) [reply Tests/Realm.Tests/Sync/SessionTests.cs:50]: Is there a specific reason for `AllClientResetHandlers` to be public?

File: Tests/Tests.iOS/Tests.iOS.csproj (hunk 9 above)
[the remark this answers is not in the document]
Comment (nirinchev) [reply Tests/Tests.iOS/Tests.iOS.csproj:34]: VS will sometimes add these automatically and it's kind of annoying.

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 10]
```diff
@@ -270,31 +237,388 @@ public void Session_ClientReset_DiscardLocal_ManualResetFallback_InitiateClientR
...
+                    if (eventArgs.PropertyName == nameof(SyncObjectWithRequiredStringList.Strings))
+                    {
+                        var list = realmA.All<SyncObjectWithRequiredStringList>().Single().Strings;
+                        Assert.That(list.Count, Is.EqualTo(3));
+                        Assert.That(list[0], Is.EqualTo("0"));
+                        Assert.That(list[1], Is.EqualTo("1"));
+                        Assert.That(list[2], Is.EqualTo("3"));
+                        tscAfterRemoteUpdateA.TrySetResult(null);
+                    }
+                };
```
[the remark this answers is not in the document]
Comment (LaPeste) [reply Tests/Realm.Tests/Sync/SessionTests.cs:531]: Then, I think we have a problem because all works fine.
You're right, I should have subscribed to `SubscribeForNotifications`, not to `PropertyChanged`.

File: Realm/Realm/Handles/SessionHandle.cs [hunk 11]
```diff
@@ -281,6 +286,7 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
                     var clientResetEx = new ClientResetException(session.User.App, messageString, errorCode, userInfo);
 
                     if (syncConfig.ClientResetHandler is DiscardLocalResetHandler ||
+                        syncConfig.ClientResetHandler is AutomaticRecoveryHandler ||
                         syncConfig.ClientResetHandler?.ManualClientReset != null)
```
Comment (nirinchev) [inline Realm/Realm/Handles/SessionHandle.cs:290]: [nit] I know that's code you didn't change, but we can remove the null propagation here
```suggestion
                        syncConfig.ClientResetHandler.ManualClientReset != null)
```

File: Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryHandler.cs [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            /// </summary>
+            Manual = 0,
+
+            /// <summary>
+            /// If the automatic merge fails, automatically discards all local changes and uses the latest realm that is available on the remote.
+            /// </summary>
+            DiscardLocal = 1,
+        }
+
+        internal Fallback FallbackStrategy { get; private set; }
```
Comment (nirinchev) [inline Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryHandler.cs:61]: We can probably expose this publicly. Not super useful, but doesn't hurt either.

File: Tests/Tests.iOS/PreserveFromStripping.xml [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<linker>
```
Comment (nirinchev) [inline Tests/Tests.iOS/PreserveFromStripping.xml:1]: We should not have this. If the managed linker strips those, we should annotate them with `[Preserve(AllMembers = true)]`.

File: Tools/DeployApps/BaasClient.cs [hunk 14]
```diff
@@ -41,8 +41,19 @@ public static class AppConfigType
...
     }
 
-    public class BaasClient : IDisposable
+    public class BaasClient
     {
+        public class FunctionReturn
+        {
+            public enum Result
+            {
+                success = 0,
```
Comment (nirinchev) [inline Tools/DeployApps/BaasClient.cs:50]: These should be documented or the warning should be suppressed.

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 15]
```diff
@@ -240,23 +210,22 @@ public void Session_ClientReset_ManualRecovery_InitiateClientReset()
...
-                    ManualResetFallback = (err) =>
-                    {
-                        manualResetFallbackHandled = true;
-                        errorTcs.TrySetResult(err);
-                    }
+                    manualResetFallbackHandled = true;
+                    errorTcs.TrySetResult(err);
                 };
 
+                config.ClientResetHandler = GetClientResetHandler(handlerType, fallback, beforeCb: null, afterCb: null, manualCb);
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:227]: `beforeCb` and `afterCb` are optional, so you can omit them.

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 16]
```diff
@@ -876,23 +1058,23 @@ public void Session_Error_OnSessionError_Coexistence()
             });
         }
 
-        [TestCase(ProgressMode.ForCurrentlyOutstandingWork)]
-        [TestCase(ProgressMode.ReportIndefinitely)]
-        public void Session_ProgressObservable_IntegrationTests(ProgressMode mode)
+        [TestCaseSource(nameof(DoubleModeClientResetHandlers))]
+        public void SessionIntegrationTest_ProgressObservable((ProgressMode Mode, ClientResetHandlerBase Handler) tuple)
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:1062]: Why change this test? Seems unrelated to the client reset work.

File: Tests/Realm.Tests/Sync/SessionTests.cs [hunk 17]
```diff
@@ -1330,6 +1524,40 @@ protected override void CustomTearDown()
 
         }
 
+        public interface IClientResetHandler
+        {
+            ClientResetHandlerBase BuildHandler(BeforeResetCallback beforeCb = null,
+                                                AfterResetCallback afterCb = null,
+                                                ClientResetCallback manualCb = null);
+        }
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SessionTests.cs:1532]: Should this be deleted

**Review by nirinchev — COMMENTED** [review]
A couple of final touches from me, but feels like it's almost ready to go.

File: Tools/DeployApps/BaasClient.cs (hunk 14 above)
Comment (LaPeste) [reply Tools/DeployApps/BaasClient.cs:50]: I missed it. I'm gonna suppress it. It's only for our internal functions results.

File: Tests/Realm.Tests/Sync/SessionTests.cs (hunk 16 above)
Comment (LaPeste) [reply Tests/Realm.Tests/Sync/SessionTests.cs:1062]: Just to keep it consistent with the other naming convention I picked for the other integration tests for the session. Is this wrong?

File: Tests/Tests.iOS/PreserveFromStripping.xml (hunk 13 above)
Comment (LaPeste) [reply Tests/Tests.iOS/PreserveFromStripping.xml:1]: @fealebenpae suggested that it's better to avoid adding that attribute when in the main assembly unless strictly necessary. In our case I get the setter of all callbacks for `AutomaticRecoveryHandler` and `DiscardLocalResetHandler` stripped. I just supposed that users should not run in the same issue. I just spotted this stripping problem because I used reflections in tests, it would otherwise have gone unnoticed.
So I'm not sure what the best course of action is here. When dealing with user facing API, should we preserve more often then not or keep it at a minimum?

Comment (nirinchev) [reply Tests/Tests.iOS/PreserveFromStripping.xml:1]: Hm... good point. I didn't realize it's due to reflection that those were stripped. In that case, let's preserve them by using them in the tests rather than with a xml file. Those can easily go out of sync with the code they're preserving and are a bit annoying to maintain (e.g. you need to also add this for MAUI and Unity). I think we can preserve them like:

```csharp
[Preserve]
static SessionTests()
{
    var preserveHandler = new AutomaticRecoveryHandler
    {
        OnBeforeReset = (...) => { },
	...
    }
}
```

File: Tools/DeployApps/BaasClient.cs (hunk 14 above)
Comment (nirinchev) [reply Tools/DeployApps/BaasClient.cs:50]: You can see the warning when you browse the code:

Screen Shot 2022-07-01 at 18 23 51

To see that in VS, you'll probably need to rebuild the project, otherwise msbuild will not show warnings for code it doesn't compile.

File: Tests/Realm.Tests/Sync/SessionTests.cs (hunk 16 above)
Comment (nirinchev) [reply Tests/Realm.Tests/Sync/SessionTests.cs:1062]: But it's changing the arguments of the test for no obvious reason. I don't see why we'd want to test different types of client reset handlers in this progress test.

Comment (LaPeste) [reply Tests/Realm.Tests/Sync/SessionTests.cs:1062]: I understand what you mean now. You're right, client resets have nothing to do here. I went over it too fast. I'll revert it to what it was before.

File: Realm/Realm/Handles/SessionHandle.cs [hunk 18]
```diff
@@ -280,8 +285,10 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
                     var userInfo = StringStringPair.UnmarshalDictionary(userInfoPairs, userInfoPairsLength.ToInt32());
                     var clientResetEx = new ClientResetException(session.User.App, messageString, errorCode, userInfo);
 
-                    if (syncConfig.ClientResetHandler is DiscardLocalResetHandler ||
-                        syncConfig.ClientResetHandler?.ManualClientReset != null)
+                    if (syncConfig.ClientResetHandler is DiscardUnsyncedChangesHandler ||
+                        syncConfig.ClientResetHandler is RecoverUnsyncedChangesHandler ||
+                        syncConfig.ClientResetHandler is RecoverOrDiscardUnsyncedChangesHandler ||
+                        syncConfig.ClientResetHandler.ManualClientReset != null)
```
Comment (nirinchev) [inline Realm/Realm/Handles/SessionHandle.cs:291]: Rather than check the types here, can we check the resync mode?

File: Realm/Realm/Handles/SessionHandle.cs [hunk 19]
```diff
@@ -323,48 +330,67 @@
...
                     var schema = syncConfig.GetSchema();
                     using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfig, schema);
-                    discardLocalHandler.OnBeforeReset(realmBefore);
+                    cb.Invoke(realmBefore);
                 }
             }
             catch (Exception ex)
             {
-                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing DiscardLocalResetHandler.OnBeforeReset during a client reset: {ex}");
+                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing {syncConfig.ClientResetHandler.GetType().Name}.{cb.GetType().Name} during a client reset: {ex}");
```
[the remark this answers is not in the document]
Comment (nirinchev) [reply Realm/Realm/Handles/SessionHandle.cs:357]: This warning is correct - we should check for null on the sync config and the callback.

File: Realm/Realm/Handles/SessionHandle.cs [hunk 20]
```diff
@@ -323,48 +330,67 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
...
```
Comment (nirinchev) [inline Realm/Realm/Handles/SessionHandle.cs:357]: Let's call this `OnBeforeReset` - there's no need to check the type of the callback.

File: Realm/Realm/Handles/SessionHandle.cs [hunk 21]
```diff
@@ -323,48 +330,67 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
...
                     using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfig, schema);
                     using var realmAfter = new Realm(new UnownedRealmHandle(after), syncConfig, schema);
-                    discardLocalHandler.OnAfterReset(realmBefore, realmAfter);
+                    cb.Invoke(realmBefore, realmAfter);
                 }
             }
             catch (Exception ex)
             {
-                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing DiscardLocalResetHandler.OnAfterReset during a client reset: {ex}");
+                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing {syncConfig.ClientResetHandler.GetType().Name}.{cb.GetType().Name} during a client reset: {ex}");
```
Comment (nirinchev) [inline Realm/Realm/Handles/SessionHandle.cs:393]: `cb.GetType().Name` is unlikely to give us anything meaningful. We should just use `OnAfterReset`.

File: Realm/Realm/Configurations/FlexibleSyncConfiguration.cs [hunk 22]
```diff
@@ -182,20 +182,22 @@ internal override Native.SyncConfiguration CreateNativeSyncConfiguration()
...
         /// <seealso href="https://docs.mongodb.com/realm/sdk/dotnet/advanced-guides/client-reset/">Client reset docs</seealso>
         public override ClientResetHandlerBase ClientResetHandler
         {
             get => base.ClientResetHandler;
             set
             {
-                if (value is DiscardLocalResetHandler)
+                if (value is DiscardUnsyncedChangesHandler ||
+                    value is RecoverUnsyncedChangesHandler ||
+                    value is RecoverOrDiscardUnsyncedChangesHandler)
```
Comment (nirinchev) [inline Realm/Realm/Configurations/FlexibleSyncConfiguration.cs:198]: It's kind of a moot point considering we're going to support all reset handler types soon, but this is not a great way to check. Apart from checking by type, it's also missing a check for `DiscardLocalResetHandler`.

File: Realm/Realm/Sync/ErrorHandling/DiscardUnsyncedChangesHandler.cs [hunk 23]
```diff
@@ -33,37 +35,52 @@ namespace Realms.Sync.ErrorHandling
...
+        /// </summary>
+        /// <value>Callback invoked if automatic Client Reset handling fails.</value>
+        public ClientResetCallback ManualResetFallback
+        {
+            get => ManualClientReset;
+            set => ManualClientReset = value;
+        }
+    }
 
+#pragma warning disable SA1402 // This type is now obsolete and will be deleted in the next major version
```
Comment (nirinchev) [inline Realm/Realm/Sync/ErrorHandling/DiscardUnsyncedChangesHandler.cs:65]:
```suggestion
#pragma warning disable SA1402 // This type is now obsolete and will be deleted in the next major version

```

File: Realm/Realm/Sync/ErrorHandling/DiscardUnsyncedChangesHandler.cs [hunk 24]
```diff
@@ -86,4 +105,5 @@ public ClientResetCallback ManualResetFallback
             set => ManualClientReset = value;
         }
     }
+#pragma warning restore SA1402 // This type is now obsolete and will be deleted in the next major version
```
Comment (nirinchev) [inline Realm/Realm/Sync/ErrorHandling/DiscardUnsyncedChangesHandler.cs:108]:
```suggestion

#pragma warning restore SA1402 // This type is now obsolete and will be deleted in the next major version

```

File: Realm/Realm/Sync/ErrorHandling/RecoverOrDiscardUnsyncedChangesHandler.cs [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+        /// <summary>
+        /// Gets or sets the callback that indicates a Client Reset is about to happen.
+        /// </summary>
+        /// <value>Callback invoked right before a Client Reset.</value>
+        public BeforeResetCallback OnBeforeReset { get; set; }
+
+        /// <summary>
+        /// Gets or sets the callback that indicates that an automatic Client Reset just happened.
+        /// </summary>
+        /// <value>Callback invoked right after a Client Reset.</value>
```
Comment (nirinchev) [inline Realm/Realm/Sync/ErrorHandling/RecoverOrDiscardUnsyncedChangesHandler.cs:63]: In the remarks we should indicate that only one of `OnAfterRecovery` or `OnAfterDiscard` but not both.

File: Tests/Realm.Tests/Sync/SyncConfigurationTests.cs [hunk 26]
```diff
@@ -109,10 +109,12 @@ public void Test_SyncConfigRelease()
         }
 
         [Test]
-        public void FlexibleSyncConfiguration_Throws_When_Assigned_DiscardLocalResetHandler()
+        public void FlexibleSyncConfiguration_WhenAssignedNewRecoveryHandlers_Throws()
         {
             var conf = GetFakeFLXConfig();
-            Assert.That(() => conf.ClientResetHandler = new DiscardLocalResetHandler(), Throws.TypeOf<NotSupportedException>());
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SyncConfigurationTests.cs:115]: This should still test the obsolete case.

File: Realm/Realm/Handles/SessionHandle.cs [hunk 27]
```diff
@@ -323,48 +330,67 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
...
+            SyncConfigurationBase syncConfig = null;
+
             try
             {
                 var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
-                var syncConfig = (SyncConfigurationBase)syncConfigHandle.Target;
+                syncConfig = (SyncConfigurationBase)syncConfigHandle.Target;
+                cb = syncConfig.ClientResetHandler switch
+                {
+                    DiscardUnsyncedChangesHandler handler => handler.OnBeforeReset,
```
Comment (nirinchev) [inline Realm/Realm/Handles/SessionHandle.cs:342]: What if it's `DiscardLocalResetHandler`? Even though the type is obsolete, we should not break users who were using it.

**Review by nirinchev — CHANGES_REQUESTED** [review]
Looks mostly fine. I've commented on several places, but be sure to audit all usages where you're explicitly checking the type of a reset handler (like `xxx is SomeResetHandler`) to make sure it also checks for `DiscardLocalResetHandler`, otherwise we're silently breaking users.

File: Realm/Realm/Handles/SessionHandle.cs (hunk 20 above)
Comment (LaPeste) [reply Realm/Realm/Handles/SessionHandle.cs:357]: All this was done to give you e.g. `DiscardUnsyncedChangesHandler.OnBeforeReset`, which I checked and it works as expected.

Comment (nirinchev) [reply Realm/Realm/Handles/SessionHandle.cs:357]: I'm not sure how it's possible to give you `DiscardUnsyncedChangesHandler.OnBeforeReset`. I haven't tried it, but I expect `cb` here to be of type `BeforeResetCallback` or `Action<Realm>`. There's no obvious reason/way for it to ever be of type `OnBeforeReset`, considering that's the name of a property and not an actual type.

Comment (LaPeste) [reply Realm/Realm/Handles/SessionHandle.cs:357]: What I said makes no sense. Without reflection this is impossible possible. And yet my memory told me "yes yes it's fine". Eh...🤦🏻

**Review by nirinchev — APPROVED** [review]
Looks good, just make sure to suppress the build warnings for the obsolete reset handler
